### PR TITLE
[XLNet] Fix mems behavior

### LINKED ISSUE #7584: XLNet evaluation fails if the size of evaluation set can't be divided by a given evaluation batch size
## Environment info

     
- `transformers` version: 3.3.1
- Platform: Linux-4.15.0-117-generic-x86_64-with-glibc2.10
- Python version: 3.8.5
- PyTorch version (GPU?): 1.4.0 (True)
- Tensorflow version (GPU?): 2.2.0 (False)
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help
@sgugger

## Information

Model I am using (Bert, XLNet ...): XLNet-base-cased

The problem arises when using:
* the official example scripts: run_glue.py

The tasks I am working on is:
* an official GLUE/SQUaD task: SST-2

## To reproduce

1. Install transformers from master and download SST-2 data using ```download_glue_data.py```

2. Create the following scripts
```bash
GLUE_DIR=~/glue
CUDA_VISIBLE_DEVICES=0
TASK_NAME=SST-2

python3 ~/applications/transformers/examples/text-classification/run_glue.py \
  --model_name_or_path ~/xlnet \
  --task_name $TASK_NAME \
  --do_eval \
  --data_dir $GLUE_DIR/$TASK_NAME \
  --max_seq_length 64 \
  --per_device_train_batch_size 32 \
  --per_device_eval_batch_size 64 \
  --learning_rate 2e-5 \
  --num_train_epochs 3.0 \
  --output_dir ~/result/$TASK_NAME/ \
  --overwrite_output_dir \
  --eval_steps 100

```
3. run this script

## Expected behavior

Trainer should return appropriate evaluation results. Here are logs when evaluating bert-base with above-given hyperparameters.
```bash
10/05/2020 22:28:47 - INFO - filelock -   Lock 140392033291808 acquired on /data/home/liusishun/glue/SST-2/cached_dev_BertTokenizer_64_sst-2.lock
10/05/2020 22:28:47 - INFO - filelock -   Lock 140392033291808 released on /data/home/liusishun/glue/SST-2/cached_dev_BertTokenizer_64_sst-2.lock
10/05/2020 22:28:50 - INFO - __main__ -   *** Evaluate ***
Evaluation: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 14/14 [00:01<00:00,  7.22it/s]
{'eval_loss': 0.6916399122378148, 'eval_acc': 0.49770642201834864, 'step': 0}
/data/home/liusishun/.conda/envs/myenv/lib/python3.8/site-packages/transformers/trainer.py:1168: FutureWarning: This method is deprecated, use `Trainer.is_world_process_zero()` instead.
  warnings.warn("This method is deprecated, use `Trainer.is_world_process_zero()` instead.", FutureWarning)
10/05/2020 22:28:52 - INFO - __main__ -   ***** Eval results sst-2 *****
10/05/2020 22:28:52 - INFO - __main__ -     eval_loss = 0.6916399122378148
10/05/2020 22:28:52 - INFO - __main__ -     eval_acc = 0.49770642201834864

```

## Observed behavior
```bash
10/05/2020 22:30:05 - INFO - filelock -   Lock 139928226197216 acquired on /data/home/liusishun/glue/SST-2/cached_dev_XLNetTokenizer_64_sst-2.lock
10/05/2020 22:30:05 - INFO - filelock -   Lock 139928226197216 released on /data/home/liusishun/glue/SST-2/cached_dev_XLNetTokenizer_64_sst-2.lock
10/05/2020 22:30:09 - INFO - __main__ -   *** Evaluate ***
Evaluation:  93%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▉           | 13/14 [00:02<00:00,  4.44it/s]
Traceback (most recent call last):
  File "/data/home/liusishun/applications/transformers/examples/text-classification/run_glue.py", line 247, in <module>
    main()
  File "/data/home/liusishun/applications/transformers/examples/text-classification/run_glue.py", line 197, in main
    eval_result = trainer.evaluate(eval_dataset=eval_dataset)
  File "/data/home/liusishun/.conda/envs/myenv/lib/python3.8/site-packages/transformers/trainer.py", line 1297, in evaluate
    output = self.prediction_loop(eval_dataloader, description="Evaluation")
  File "/data/home/liusishun/.conda/envs/myenv/lib/python3.8/site-packages/transformers/trainer.py", line 1382, in prediction_loop
    preds = logits if preds is None else nested_concat(preds, logits, dim=0)
  File "/data/home/liusishun/.conda/envs/myenv/lib/python3.8/site-packages/transformers/trainer_utils.py", line 151, in nested_concat
    return type(tensors)(nested_concat(t, n, dim) for t, n in zip(tensors, new_tensors))
  File "/data/home/liusishun/.conda/envs/myenv/lib/python3.8/site-packages/transformers/trainer_utils.py", line 151, in <genexpr>
    return type(tensors)(nested_concat(t, n, dim) for t, n in zip(tensors, new_tensors))
  File "/data/home/liusishun/.conda/envs/myenv/lib/python3.8/site-packages/transformers/trainer_utils.py", line 151, in nested_concat
    return type(tensors)(nested_concat(t, n, dim) for t, n in zip(tensors, new_tensors))
  File "/data/home/liusishun/.conda/envs/myenv/lib/python3.8/site-packages/transformers/trainer_utils.py", line 151, in <genexpr>
    return type(tensors)(nested_concat(t, n, dim) for t, n in zip(tensors, new_tensors))
  File "/data/home/liusishun/.conda/envs/myenv/lib/python3.8/site-packages/transformers/trainer_utils.py", line 152, in nested_concat
    return torch.cat((tensors, new_tensors), dim=dim)
RuntimeError: invalid argument 0: Sizes of tensors must match except in dimension 0. Got 40 and 64 in dimension 1 at /opt/conda/conda-bld/pytorch_1579061855666/work/aten/src/THC/generic/THCTensorMath.cu:71
```

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Fixes #7584 

XLNet is arguably not the cleanent implementation with a lot of config parameters flying around that all interact with each other in an overly complicated way: `use_cache`, `mem_len`, and `reuse_len`. Due to backward compatibility we cannot really get rid of those. I'd love to just completely get rid of `use_cache` and `reuse_len`, but this would require to update all configs which is not possible...

At first, this PR removes the `use_cache` and replaces it with `use_mems_eval` => `use_mems_eval` decides whether the mems should be used in evaluation mode, which defaults to `True` so that the arguably most important model `XLNetLMHeadModel` keeps full backward compatibility at inference. `use_cache` is a confusing name IMO because it does not correspond to the `use_cache` we know from GPT2 (we had a longer discussion on this internally on Slack).

The issue #7584 shows that if there is one training batch that is smaller than the other batches in the train data, the training breaks. Also as can be read upon in the issue linked below, the authors also don't use the memory mechanism for fine-tuning => therefore we add another param `use_mems_train` which defaults to `False` so that training works as a default. 
If for some special reason the user wants to use the memory mechanism during fine-tuning, he/she has to make sure that the batch_size of all training batches is the same.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to the it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

File: src/transformers/modeling_xlnet.py [hunk 1]
```diff
@@ -1091,7 +1091,11 @@ def forward(
             output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
         )
         return_dict = return_dict if return_dict is not None else self.config.use_return_dict
-        use_cache = self.training or (use_cache if use_cache is not None else self.config.use_cache)
+
+        if self.training:
+            use_mems = use_mems if use_mems is not None else self.config.use_mems_train
+        else:
+            use_mems = use_mems if use_mems is not None else self.config.use_mems_eval
```
Comment (sgugger) [inline src/transformers/modeling_xlnet.py:1098]: Nit: I'd personally prefer in two lines:
```
default_use_mems = self.config.use_mems_train if self.training else self.config.use_mems_eval
use_mems = use_mems if use_mems is not None else default_use_mems
```
but it's a matter of personal taste, probably.

**Review by sgugger — APPROVED** [review]
Thank you so much for digging into this!!!

**Review by LysandreJik — COMMENTED** [review]
It is great that you deep dived into this issue and solved it, thank you! It seems like we're losing backwards compatibility on the `use_cache` parameter.

The new arguments are perfectly acceptable, but I think it would be nice to keep backwards compatibility. It should be easy enough by checking if there's a `use_cache` value in the configuration and setting it to the value of `use_mems`.

The following code does not output equal results on `master` and on this PR:

```py
from transformers import XLNetModel, XLNetTokenizer

model = XLNetModel.from_pretrained("xlnet-base-cased", use_cache=False)
tokenizer = XLNetTokenizer.from_pretrained("xlnet-base-cased")
inputs = tokenizer("Nice isn't it", return_tensors="pt")
outputs = model(**inputs)
assert len(outputs) == 1
```
The assertion passes on `master`, but not on this branch.

**Comment by patrickvonplaten** [comment]
> The new arguments are perfectly acceptable, but I think it would be nice to keep backwards compatibility. It should be easy enough by checking if there's a `use_cache` value in the configuration and setting it to the value of `use_mems`.

Yes, true! The problem is that `use_cache` is always in `configuration_utils.py` so we can't really add something that checks whether `use_cache` is in the config (It's always there) and sets `use_mems` to the same value (It'd always do it)...

Thinking more about it `use_cache` should maybe not even be in `configuration_utils.py` but just in all "LMHead" configs?! I could move `use_cache` to all individual config files. Wdyt? @LysandreJik @sgugger ?

**Comment by sgugger** [comment]
I have mixed feelings about this, on one hand it makes sense to remove that argument which is very specific to a few models from the main `configuration_utils`, but on the other I fear the associated breaking changes.

**Comment by LysandreJik** [comment]
I think your last proposal would be great @patrickvonplaten, moving `use_cache` to the appropriate configuration files.

File: src/transformers/models/xlnet/configuration_xlnet.py [hunk 2]
```diff
@@ -197,6 +207,16 @@ def __init__(
         self.pad_token_id = pad_token_id
         self.eos_token_id = eos_token_id
 
+        if "use_cache" in kwargs:
```
Comment (patrickvonplaten) [inline src/transformers/models/xlnet/configuration_xlnet.py:210]: this ensures backward compatibility.

**Comment by patrickvonplaten** [comment]
## Update 

@LysandreJik @sgugger - `use_cache` is only used in the following modeling files (TF & PT):
```
- modeling_t5.py
- modeling_bart.py (+ inherited)
- modeling_fstm.py
- modeling_prophetnet.py
- modeling_gpt2.py
- modeling_openai.py
- modeling_xlnet.py
- modeling_ctrl.py
```

Therefore, we can move `use_cache` to the respective configuration files and delete it from the general `configuration_utils.py` file.
I cannot really think of a use case where this would lead to breaking changes. If *e.g.* a saved BERT config includes a `use_cache`, this parameter will still be set to the config: https://github.com/huggingface/transformers/blob/18c8cf000bed04ec03470270ec2fbd9d49cce5c4/src/transformers/configuration_utils.py#L235 and still remain unused as before => I don't think this is a problem. A lot of old configs have unused config parameters that are then just ignored... For all models that make use of `use_cache`, the `use_cache` param is now "caught" and saved directly in the model's config. 

=> this allows us to have 0 breaking changes for this PR and is also cleaner IMO.

File: src/transformers/models/xlnet/configuration_xlnet.py [hunk 3]
```diff
@@ -106,12 +108,18 @@ class XLNetConfig(PretrainedConfig):
...
-            Whether or not the model should return the last pre-computed hidden states.
+        use_mems_eval (:obj:`bool`, `optional`, defaults to :obj:`True`):
+            Whether or not the model should make use of the recurrent memory mechanism in evaluation mode.
+        use_mems_train (:obj:`bool`, `optional`, defaults to :obj:`False`):
+            Whether or not the model should make use of the recurrent memory mechanism in train mode.
 
             .. note::
-                This flag behaves differently from with other models: it just controls the inference behavior, during
-                training the model always uses ``use_cache=True``.
+                For pre-training, it is recommended to set ``use_mems_train`` to :obj:`True`. For fine-tuning, it is
```
Comment (sgugger) [inline src/transformers/models/xlnet/configuration_xlnet.py:117]:
```suggestion
                For pretraining, it is recommended to set ``use_mems_train`` to :obj:`True`. For fine-tuning, it is
```

**Review by sgugger — APPROVED** [review]
This LGTM and it's way cleaner this way, thanks a lot!

File: src/transformers/models/xlnet/configuration_xlnet.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/transformers/models/xlnet/configuration_xlnet.py]: I think this happens quite often :D Grammarly tells me `pre-training` is better - but I'm 100% fine with changing my spelling to `pretraining`. I'll replace all pre-training's with pretrainings in `src/transformers` and `docs` :-)

**Review by LysandreJik — APPROVED** [review]
Ok, this LGTM. Should we include this in the v4.0.0 @patrickvonplaten ?

**Comment by patrickvonplaten** [comment]
## Update

To have a bit more clarity for this PR.
This PR mainly solved the issue that XLNet cannot be trained at the moment. It does so by depreciating `use_cache` and replacing it by `use_mems_eval` and `use_mems_train`.
The PR keeps 100% backward compatibility for the PyTorch model.

The TF XLNet model was previously not kept up-to-date with the PT model. This PR notably forgot to update the TF model: https://github.com/huggingface/transformers/pull/5770 when the behavior of the PT model was changed. 
This TF brings the TF model up to date with the PT model, but does not keep 100% backward compatibility by completely removing the `use_cache` parameter. I decided to not add the `use_cache` parameter because a) Bringing the TF model up to date with PT is breaking anyways and b) there are breaking changes anyways for boolean inputs in the moment in TF. So in conclusion:

- for PT: No breaking changes -> `use_cache` is depreciated both in the config and in the model forward's
- for TF: One breaking change for TF that `use_cache` cannot be forwarded anymore to the models and has to be replaced by `use_mems`

File: src/transformers/models/xlnet/modeling_tf_xlnet.py [hunk 4]
```diff
@@ -569,7 +581,7 @@ def call(
         input_mask=None,
         head_mask=None,
         inputs_embeds=None,
-        use_cache=True,
```
Comment (patrickvonplaten) [inline src/transformers/models/xlnet/modeling_tf_xlnet.py:572]: this was very out-of-date `use_cache` should not default to `True` as a function input -> it makes `config.use_cache` useless.

File: src/transformers/models/xlnet/modeling_tf_xlnet.py [hunk 5]
```diff
@@ -489,14 +492,23 @@ def create_mask(self, qlen, mlen, dtype=tf.float32):
         return ret
 
     def cache_mem(self, curr_out, prev_mem):
-        """cache hidden states into memory."""
+        # cache hidden states into memory.
         if self.reuse_len is not None and self.reuse_len > 0:
             curr_out = curr_out[: self.reuse_len]
 
+        if self.mem_len is None or self.mem_len == 0:
```
Comment (patrickvonplaten) [inline src/transformers/models/xlnet/modeling_tf_xlnet.py:499]: this aligns the TF behavior with the PT behavior

File: src/transformers/models/xlnet/modeling_xlnet.py [hunk 6]
```diff
@@ -1080,18 +1081,30 @@ def forward(
         input_mask=None,
         head_mask=None,
         inputs_embeds=None,
-        use_cache=None,
+        use_mems=None,
         output_attentions=None,
         output_hidden_states=None,
         return_dict=None,
+        **kwargs,  # delete after depreciation warning is removed
```
Comment (patrickvonplaten) [inline src/transformers/models/xlnet/modeling_xlnet.py:1088]: keep backward compatibility in PT for now

**Comment by LysandreJik** [comment]
Thank you for clarifying, I think that's fine. We should add that to the v4.0.0 as well. 

Thanks for taking care of it @patrickvonplaten

**Pull request merged**